### ISSUE TITLE
MDQ broken with MySQL 5.7

### ISSUE THREAD

**Issue by vladimir-mencl-eresearch** [ausaccessfed/saml-service#166]
Hi @smangelsdorf 

I've just merged the latest develop to my saml service DEV copy and I can no longer access the MDQ URL.

When I try:

    wget -O tuakiri-dev-entities.xml x --header 'Accept: application/samlmetadata+xml' https://saml.dev.tuakiri.ac.nz/mdq/dev-tuakiri-ac-nz/entities

I get a 500 Internal Server error.  The Ruby error page says:

    Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column &#39;service_names.value&#39; which is not in SELECT list; this is incompatible with DISTINCT

After merging the changes, I did run

    gem install bundler
    bundler install
    rake db:migrate

So I'm assuming I have the right table structure.

I don't know if this is a bug in the new code or misconfiguration at my end - but reporting it in case it's helpful.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
The relevant bits from what I see in the logs are:

```
  Sequel::Mysql2::Database (0.8ms)  SELECT `endpoints`.`id`, `endpoints`.`binding`, `endpoints`.`location`, `endpoints`.`response_location`, `endpoints`.`created_at`, `endpoints`.`updated_at`, `indexed_endpoints`.`is_default`, `indexed_endpoints`.`index`, `assertion_consumer_services`.`sp_sso_descriptor_id` FROM `endpoints` INNER JOIN `indexed_endpoints` ON (`indexed_endpoints`.`id` = `endpoints`.`id`) INNER JOIN `assertion_consumer_services` ON (`assertion_consumer_services`.`id` = `indexed_endpoints`.`id`) WHERE (`assertion_consumer_services`.`sp_sso_descriptor_id` = 68) ORDER BY `index`, `location`

==> log/unicorn-stderr.log <==
Exception `Mysql2::Error' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/mysql2-0.4.6/lib/mysql2/client.rb:120 - Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT
Exception `Mysql2::Error' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/activesupport-5.1.1/lib/active_support/notifications/instrumenter.rb:25 - Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT

==> log/development.log <==
  Sequel::Mysql2::Database (1.4ms)  SELECT DISTINCT `attribute_consuming_services`.* FROM `attribute_consuming_services` INNER JOIN (SELECT `localized_names`.`id`, `localized_names`.`value`, `localized_names`.`lang`, `localized_names`.`created_at`, `localized_names`.`updated_at`, `service_names`.`attribute_consuming_service_id` FROM `localized_names` INNER JOIN `service_names` ON (`service_names`.`id` = `localized_names`.`id`)) AS `service_names` ON (`service_names`.`attribute_consuming_service_id` = `attribute_consuming_services`.`id`) WHERE (`attribute_consuming_services`.`sp_sso_descriptor_id` = 68) ORDER BY `service_names`.`value`

==> log/unicorn-stderr.log <==
Exception `Mysql2::Error' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/sequel-rails-0.9.15/lib/sequel_rails/sequel/database/active_support_notification.rb:21 - Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT
Exception `Sequel::DatabaseError' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/sequel-4.47.0/lib/sequel/database/misc.rb:421 - Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT

==> log/development.log <==
  Sequel::Mysql2::Database (0.4ms)  ROLLBACK

==> log/unicorn-stderr.log <==
Exception `Sequel::DatabaseError' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/sequel-4.47.0/lib/sequel/database/misc.rb:423 - Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT
Exception `Sequel::DatabaseError' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/actionpack-5.1.1/lib/action_controller/metal/rescue.rb:23 - Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT
Exception `Sequel::DatabaseError' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/activesupport-5.1.1/lib/active_support/notifications/instrumenter.rb:25 - Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT

==> log/development.log <==
Completed 500 Internal Server Error in 1270ms (Models: 597.3ms)



==> log/unicorn-stderr.log <==
Exception `Sequel::DatabaseError' at /home/samlsvc/.rbenv/versions/2.2.7/lib/ruby/gems/2.2.0/gems/actionpack-5.1.1/lib/action_dispatch/middleware/callbacks.rb:30 - Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT

==> log/development.log <==

Sequel::DatabaseError (Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT):

lib/metadata/saml.rb:486:in `block in sp_sso_descriptor'
lib/metadata/saml.rb:480:in `sp_sso_descriptor'
lib/metadata/saml.rb:265:in `block (2 levels) in entity_descriptor'
lib/metadata/saml.rb:264:in `each'
lib/metadata/saml.rb:264:in `block in entity_descriptor'
lib/metadata/saml.rb:256:in `entity_descriptor'
lib/metadata/saml.rb:71:in `block in render_known_entities'
lib/metadata/saml.rb:70:in `each'
lib/metadata/saml.rb:70:in `render_known_entities'
lib/metadata/saml.rb:64:in `block (2 levels) in entities_descriptor'
lib/metadata/saml.rb:60:in `block in entities_descriptor'
lib/metadata/saml.rb:55:in `entities_descriptor'
app/controllers/concerns/metadata_query_caching.rb:59:in `block in cache_known_entities_response'
app/controllers/concerns/metadata_query_caching.rb:58:in `cache_known_entities_response'
app/controllers/metadata_query_controller.rb:122:in `create_known_entities_response'
app/controllers/metadata_query_controller.rb:56:in `block in handle_entities_request'
app/controllers/metadata_query_controller.rb:47:in `handle_entities_request'
app/controllers/metadata_query_controller.rb:21:in `all_entities'
```

**Comment by vladimir-mencl-eresearch** [comment]
And adding the Rails error page in case it helps...

[err.zip](https://github.com/ausaccessfed/saml-service/files/1123545/err.zip)

**Comment by bradleybeddoes** [comment]
Which MySQL version is this? I've not seen this behaviour.

**Comment by vladimir-mencl-eresearch** [comment]
Hi @bradleybeddoes , sorry, this question slipped in my inbox.

The DB used was MySQL 5.7.22-0ubuntu0.16.04.1 onto Ubuntu 16.04.

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
@bradleybeddoes , btw I've just checked with latest `develop` and still getting the same:

```
Sequel::DatabaseError in MetadataQueryController#all_entities
Mysql2::Error: Expression #1 of
```

```
lib/metadata/saml.rb:503:in `block in sp_sso_descriptor'
lib/metadata/saml.rb:497:in `sp_sso_descriptor'
lib/metadata/saml.rb:256:in `block (2 levels) in entity_descriptor'
lib/metadata/saml.rb:255:in `each'
lib/metadata/saml.rb:255:in `block in entity_descriptor'
lib/metadata/saml.rb:247:in `entity_descriptor'
lib/metadata/saml.rb:71:in `block in render_known_entities'
lib/metadata/saml.rb:70:in `each'
lib/metadata/saml.rb:70:in `render_known_entities'
lib/metadata/saml.rb:64:in `block (2 levels) in entities_descriptor'
lib/metadata/saml.rb:60:in `block in entities_descriptor'
lib/metadata/saml.rb:55:in `entities_descriptor'
app/controllers/concerns/metadata_query_caching.rb:59:in `block in cache_known_entities_response'
app/controllers/concerns/metadata_query_caching.rb:58:in `cache_known_entities_response'
app/controllers/metadata_query_controller.rb:122:in `create_known_entities_response'
app/controllers/metadata_query_controller.rb:56:in `block in handle_entities_request'
app/controllers/metadata_query_controller.rb:47:in `handle_entities_request'
app/controllers/metadata_query_controller.rb:21:in `all_entities'
```

**Comment by bradleybeddoes** [comment]
Can't say I can trigger it on this side, looks like something with MySQL 5.7, we're on 5.6.

```
$ wget -O entities.xml --header 'Accept: application/samlmetadata+xml' https://saml.test.aaf.edu.au/mdq/aaf/entities
--2018-05-08 13:33:22--  https://saml.test.aaf.edu.au/mdq/aaf/entities
Resolving saml.test.aaf.edu.au (saml.test.aaf.edu.au)... 52.64.51.21
Connecting to saml.test.aaf.edu.au (saml.test.aaf.edu.au)|52.64.51.21|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/samlmetadata+xml]
Saving to: ‘entities.xml’

entities.xml                                    [  <=>                                                                                    ]   3.54M  15.9MB/s    in 0.2s

2018-05-08 13:34:14 (15.9 MB/s) - ‘entities.xml’ saved [3717494]


$ head entities.xml
<?xml version="1.0" encoding="UTF-8"?>
<EntitiesDescriptor xmlns="urn:oasis:names:tc:SAML:2.0:metadata" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:idpdisc="urn:oasis:names:tc:SAML:profiles:SSO:idp-discovery-protocol" xmlns:mdrpi="urn:oasis:names:tc:SAML:metadata:rpi" xmlns:mdui="urn:oasis:names:tc:SAML:metadata:ui" xmlns:mdattr="urn:oasis:names:tc:SAML:metadata:attribute" xmlns:shibmd="urn:mace:shibboleth:metadata:1.0" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:fed="http://docs.oasis-open.org/wsfed/federation/200706" xmlns:privacy="http://docs.oasis-open.org/wsfed/privacy/200706" xmlns:remd="http://refeds.org/metadata" ID="aaf20180508033322" Name="https://md.test.aaf.edu.au/aaf-test-metadata.xml" validUntil="2018-05-15T03:33:22Z">
  <ds:Signature>
    <ds:SignedInfo>
      <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
      <ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
      <ds:Reference URI="#aaf20180508033322">
        <ds:Transforms>
          <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
          <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>

...
```

**Comment by vladimir-mencl-eresearch** [comment]
Hi @bradleybeddoes 

With a thinking cap on, I've managed to narrow this down to one particular EntityDescriptor, for an SP that does not request any attributes in FR and so does not have an AttributeConsumingService element at all - and so also no ServiceName.

Same error when coming via
```
https://saml.dev.tuakiri.ac.nz/mdq/dev-tuakiri-ac-nz/entities/%7Bsha1%7D6128a70c13f11b63363350a8df5e17e926e32a52
```

Does this help?

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
Attaching:
* metadata of the troublesome EntityDescriptor
* error page rendered when accessing the single entity

[err-single-entity.zip](https://github.com/ausaccessfed/saml-service/files/1985735/err-single-entity.zip)

**Comment by vladimir-mencl-eresearch** [comment]
Hi @bradleybeddoes ,

Apologies, I got confused on this and I need to correct my position:
* I can correctly render metadata for entities that do NOT contain an SPSSODescriptor.
* I however get the failure for entities that contain an SPSSODescriptor, regardless of whether they contain an AttributeConsumingService or not.
(Apologies, I got confused about this by running a test against a wrong entityID - the IdP side of the VHO instead of the SP side of the VHO).

I'm now looking again at the MySQL code that trips the error:
```
SELECT DISTINCT `attribute_consuming_services`.* FROM `attribute_consuming_services` INNER JOIN (SELECT `localized_names`.`id`, `localized_names`.`value`, `localized_names`.`lang`, `localized_names`.`created_at`, `localized_names`.`updated_at`, `service_names`.`attribute_consuming_service_id` FROM `localized_names` INNER JOIN `service_names` ON (`service_names`.`id` = `localized_names`.`id`)) AS `service_names` ON (`service_names`.`attribute_consuming_service_id` = `attribute_consuming_services`.`id`) WHERE (`attribute_consuming_services`.`sp_sso_descriptor_id` = 150) ORDER BY `service_names`.`value`

Sequel::DatabaseError (Mysql2::Error: Expression #1 of ORDER BY clause is not in SELECT list, references column 'service_names.value' which is not in SELECT list; this is incompatible with DISTINCT):
```

The SQL is a bit tricky to navigate through (double nested INNER JOIN), but, at the top level, it starts with `SELECT ... attribute_consuming_services.* FROM...`

It does NOT list the columns from the joined tables to be selected - and MySQL is complaining about that - or, specifically, about using one of these columns for ORDER BY.

If I change that SQL to also explicitly include the `service_names.value` field, it works:
```
SELECT DISTINCT `attribute_consuming_services`.*, service_names.value FROM `attribute_consuming_services` INNER JOIN (SELECT `localized_names`.`id`, `localized_names`.`value`, `localized_names`.`lang`, `localized_names`.`created_at`, `localized_names`.`updated_at`, `service_names`.`attribute_consuming_service_id` FROM `localized_names` INNER JOIN `service_names` ON (`service_names`.`id` = `localized_names`.`id`)) AS `service_names` ON (`service_names`.`attribute_consuming_service_id` = `attribute_consuming_services`.`id`) WHERE (`attribute_consuming_services`.`sp_sso_descriptor_id` = 150) ORDER BY `service_names`.`value`;
```

I have no idea how this relates to the way Rails constructs the SQL code (and what would be the way to fix it) - but at the SQL level, this is what makes the difference at my end....

Hope this helps.

Cheers,
Vlad

**Comment by bradleybeddoes** [comment]
Thanks - will leave this open for if/when we consider a move to MySQL 5.7, in the interim if you have a pressing need it will be the version of [Sequel](https://sequel.jeremyevans.net/) that we're using that will need to be investigated and tweaked here.

**Comment by vladimir-mencl-eresearch** [comment]
Thanks @bradleybeddoes .

I've just tried upgrading sequel from 4.49.0 to 5.8.0 and ended up getting:
```
Sequel::Error (class_table_inheritance with duplicate column names (other than the primary key column) is not supported, make sure tables have unique column names (duplicate columns: [:entity_descriptor_id]). If this is desired, specify these columns in the :ignore_subclass_columns option when initializing the plugin):
```

So, the upgrade won't be as easy ... I will now park this at my end too, and if we get to actually roll out SAML service, I'll dig into with the above information...

Thanks for all your help with this!

Cheers,
Vlad

**Comment by vladimir-mencl-eresearch** [comment]
Fixed by  f99eaf4 in #192
